src/mqtt_interface.rs

```diff
@@ -0,0 +1,258 @@
+        // Note(unwrap): We maintain strict control of the client object, so it should always be
+        // present.
+        let mut client = self.client.take().unwrap();
+
+        // If we are not yet subscribed to the necessary topics, subscribe now.
+        if !self.subscribed && client.is_connected()? {
+            client.subscribe(&self.settings_topic, &[])?;
+            client.subscribe(&self.commit_topic, &[])?;
+            self.subscribed = true;
+        }
+
+        let mut result = Action::Continue;
+
+        let result = match client.poll(|client, topic, message, properties| {
+            let mut split = topic.split('/');
+
+            // Publish the response to the request over MQTT using the ResponseTopic property if
+            // possible. Otherwise, default to a logging topic.
+            let response_topic = if let Some(Property::ResponseTopic(topic)) =
+                properties.iter().find(|&prop| {
+                    if let Property::ResponseTopic(_) = *prop {
+                        true
+                    } else {
+                        false
+                    }
+                }) {
+                *topic
+            } else {
+                &self.default_response_topic
+            };
+
+            // Verify topic ID against our ID.
+            let id = split.next();
+            if id.is_none() {
+                // Make a best-effort attempt to send the response. If we get a failure, we may have
+                // disconnected or the peer provided an invalid topic to respond to. Ignore the
+                // failure in these cases.
+                client
+                    .publish(
+                        response_topic,
+                        "No ID speciifed".as_bytes(),
+                        QoS::AtMostOnce,
+                        &[],
+                    )
+                    .ok();
+                return;
+            }
+
+            if id.unwrap() != self.id {
+                let mut response: String<consts::U512> = String::new();
+                write!(&mut response, "Invalid ID: {:?}", id)
+                    .unwrap_or_else(|_| response = String::from("Bad ID"));
+
+                // Make a best-effort attempt to send the response. If we get a failure, we may have
+                // disconnected or the peer provided an invalid topic to respond to. Ignore the
+                // failure in these cases.
+                client
+                    .publish(response_topic, &response.into_bytes(), QoS::AtMostOnce, &[])
+                    .ok();
+                return;
+            }
+
+            // Process the command
+            let response = match split.next() {
+                Some("settings") => {
+                    // Handle settings failures
+                    let mut response: String<consts::U512> = String::new();
+                    match self.settings.string_set(split.peekable(), message) {
```

Comment:
This library is doing some pretty large stack allocations between these temporary buffers and recursive calls to `string_set()`, I wonder if we should do some analysis and mention it in the README? 

It's probably fine for RTIC with its "one stack" design, but threaded RTOSs can often have pretty small stacks by default, 1KB or even less.

Reply:
Right. Stack usage here and in minimq in addition to the memory usage in the nal and smoltcp sounds like it might be a bit on the high side.

Reply:
I've refactored this to take significantly less stack. Note that the recursive calls shouldn't be adding very much overhead in each call, since they're extremely simple functions. Additionally, there's not much overhead using this as a layer between MQTT and the application, as it's only a single function call of abstraction. The network servicing routines are not called within these scopes, so that's not of concern here.